### DISCUSSION TITLE
Including all nested ParseObjects two levels deep.

### DISCUSSION THREAD

**Discussion by vdkdamian** [netreconlab/Parse-Swift#31] [Q&A] [answered]
I have a database with some pointers. But when using `.includeAll`, I only get nested ParseObjects from one level deep.
Is there a way to get them two levels deep?

**Comment by cbaker6** [comment, marked as answer]
This is expected behavior and one level deep is listed in the [documentation](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/query/includeall()). The server currently only allows 1-deep. You can see my comments and discussion about this with relevant links here: https://github.com/parse-community/Parse-Swift/issues/359#issuecomment-1119198294

If you want to go multiple layers deep, i.e 3-deep, and you know where you want to stop, use:

```swift
// Each of these are a Parse Pointer
.include("1deep.2deep.3deep", "field2", etc.)
```

Note, if the server ever enables this type of feature, there's a potential for infinite loops as the client SDK's are what has protected against cycles historically, not the server. All clients have their own specific designs for detecting cycles (for example, the Swift SDK is nothing like the others), so if a client SDK introduced a cycle on your server via a bug that has never been patched or your app allows data to be uploaded directly through REST or GraphQL and skipped a cycle check, you are in trouble. In addition, I can see a number of different nightmares introduced by this feature  as well... I recommend sticking with specific queries (like my example above) and when needed querying again if you want your system to remain efficient.

**Reply by vdkdamian** [reply]
I just tried it trying this: `.include("1deep.2deep", "shop", "product", "category", "shopType")`, but this doesn't seem to work.

**Reply by cbaker6** [reply]
If it's not working, you probably need to update your server to 5+ (from your [previous post](https://github.com/netreconlab/Parse-Swift/issues/13#issue-1438557098), you are on 4.x). I fixed issues on the server related to how the Parse Server digests JSON which was released on 5.0.0, https://github.com/parse-community/parse-server/pull/7242. If that doesn't work, I recommend opening an issue on the server.

**Reply by cbaker6** [reply]
You can also try 1-deep first as this is the furthest test I found on the server while doing a quick browse: https://github.com/parse-community/parse-server/blob/fded5be186952f8aab273096ba249ab82d7177bb/spec/ParseQuery.spec.js#L2584-L2617

If 1-deep works, but n-deep doesn't, then I recommend opening an issue on the server. They will most likely ask you to copy/edit the test above for n-deep to see if it works.

**Reply by cbaker6** [reply]
I know [select](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/query/select(_:)-23ewt) and [exclude](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/query/exclude(_:)-1jvpb) works for n-level deep which `select` should also give you what you need, but you will definitely need server 5+ as it requires my server fixes above as noted in the documentation. There are tests for n-level deep for those as well as combinations of `include/select`: https://github.com/parse-community/parse-server/blob/fded5be186952f8aab273096ba249ab82d7177bb/spec/ParseQuery.spec.js#L4604-L4658

**Reply by vdkdamian** [reply]
Yes indeed, not on 5+ yet. It's not supported yet by Back4App. I reached out to them. I'll try later to check if it's working. Thanks

**Reply by vdkdamian** [reply]
I'm now on 5.2.3, i can't get it working.

**Reply by cbaker6** [reply]
You must be going something incorrectly in your code as I modified the following [Playground](https://github.com/netreconlab/Parse-Swift/blob/main/ParseSwift.playground/Pages/8%20-%20Pointers.xcplaygroundpage/Contents.swift) with the following code:

```swift
struct Page: ParseObject {
    //: These are required by ParseObject
    var objectId: String?
    var createdAt: Date?
    var updatedAt: Date?
    var ACL: ParseACL?
    var score: Double?
    var originalData: Data?
    
    var words: String?
}

//: Create your own value typed `ParseObject`.
struct Book: ParseObject, ParseQueryScorable {
    //: These are required by ParseObject
    var objectId: String?
    var createdAt: Date?
    var updatedAt: Date?
    var ACL: ParseACL?
    var score: Double?
    var originalData: Data?

    //: Your own properties.
    var title: String?
    var relatedBook: Pointer<Book>?
    var page: Page?

    /*:
     Optional - implement your own version of merge
     for faster decoding after updating your `ParseObject`.
     */
    func merge(with object: Self) throws -> Self {
        var updated = try mergeParse(with: object)
        if updated.shouldRestoreKey(\.title,
                                     original: object) {
            updated.title = object.title
        }
        if updated.shouldRestoreKey(\.relatedBook,
                                     original: object) {
            updated.relatedBook = object.relatedBook
        }
        return updated
    }
}

/*:
 It's recommended to place custom initializers in an extension
 to preserve the memberwise initializer.
 */
extension Book {

    init(title: String) {
        self.title = title
    }
}

struct Author: ParseObject {
    //: These are required by ParseObject.
    var objectId: String?
    var createdAt: Date?
    var updatedAt: Date?
    var ACL: ParseACL?
    var originalData: Data?

    //: Your own properties.
    var name: String?
    var book: Book?
    var otherBooks: [Book]?

    /*:
     Optional - implement your own version of merge
     for faster decoding after updating your `ParseObject`.
     */
    func merge(with object: Self) throws -> Self {
        var updated = try mergeParse(with: object)
        if updated.shouldRestoreKey(\.name,
                                     original: object) {
            updated.name = object.name
        }
        if updated.shouldRestoreKey(\.book,
                                     original: object) {
            updated.book = object.book
        }
        if updated.shouldRestoreKey(\.otherBooks,
                                     original: object) {
            updated.otherBooks = object.otherBooks
        }
        return updated
    }
}

/*:
 It's recommended to place custom initializers in an extension
 to preserve the memberwise initializer.
 */
extension Author {
    init(name: String, book: Book) {
        self.name = name
        self.book = book
    }
}

var newBook = Book(title: "hello")
var page = Page()
page.words = "Hello world"
newBook.page = page
let author = Author(name: "Alice", book: newBook)

author.save { result in
    switch result {
    case .success(let savedAuthorAndBook):
        assert(savedAuthorAndBook.objectId != nil)
        assert(savedAuthorAndBook.createdAt != nil)
        assert(savedAuthorAndBook.updatedAt != nil)

        print("Saved: \(savedAuthorAndBook)")
    case .failure(let error):
        assertionFailure("Error saving: \(error)")
    }
}

//: Pointer array.
let otherBook1 = Book(title: "I like this book")
let otherBook2 = Book(title: "I like this book also")
var author2 = Author(name: "Bruce", book: newBook)
author2.otherBooks = [otherBook1, otherBook2]

author2.save { result in
    switch result {
    case .success(let savedAuthorAndBook):
        assert(savedAuthorAndBook.objectId != nil)
        assert(savedAuthorAndBook.createdAt != nil)
        assert(savedAuthorAndBook.updatedAt != nil)
        assert(savedAuthorAndBook.otherBooks?.count == 2)

        /*:
         Notice the pointer objects have not been updated on the
         client.If you want the latest pointer objects, fetch and include them.
         */
        print("Saved: \(savedAuthorAndBook)")

    case .failure(let error):
        assertionFailure("Error saving: \(error)")
    }
}

//: Query for your new saved author
let query1 = Author.query("name" == "Bruce")
    .include("book.page", "otherBooks")

query1.first { results in
    switch results {
    case .success(let author):
        print("Found author: \(author)")

    case .failure(let error):
        assertionFailure("Error querying: \(error)")
    }
}
```

And it returned the expected results with `Page` which is 2 levels deep:

```js
Found author: Author ({"book":{"createdAt":{"__type":"Date","iso":"2022-12-14T20:22:51.542Z"},"objectId":"vuPlNPQmt1OJdhWbYTTAtA4we8DLl4xS","page":{"createdAt":{"__type":"Date","iso":"2022-12-14T20:22:51.485Z"},"objectId":"SWdsxBfRQOVVH9TPXI5jNLHjlkzqymvd","updatedAt":{"__type":"Date","iso":"2022-12-14T20:22:51.485Z"},"words":"Hello world"},"title":"hello","updatedAt":{"__type":"Date","iso":"2022-12-14T20:22:51.542Z"}},"createdAt":{"__type":"Date","iso":"2022-12-14T20:22:51.555Z"},"name":"Bruce","objectId":"7NR1Sfmk4IP9xNIiOZ2zakt6ihtWRwJI","otherBooks":[{"createdAt":{"__type":"Date","iso":"2022-12-14T20:22:51.486Z"},"objectId":"iNyvBSPw3HrqhPCb87pyAe47z59npSx2","title":"I like this book","updatedAt":{"__type":"Date","iso":"2022-12-14T20:22:51.486Z"}},{"createdAt":{"__type":"Date","iso":"2022-12-14T20:22:51.486Z"},"objectId":"xj6iI3wUePrjFj40XnKK2PSj17marxKL","title":"I like this book also","updatedAt":{"__type":"Date","iso":"2022-12-14T20:22:51.486Z"}}],"updatedAt":{"__type":"Date","iso":"2022-12-14T20:22:51.555Z"}})
```

**Reply by vdkdamian** [reply]
Thanks, your example cleared up a lot! It's working fine now
